spec/case-selection.md

```diff
@@ -0,0 +1,68 @@
+## Variant Case Selection
+
+> This is a fragment of the full MessageFormat 2 specification,
+> intended for later merge into a complete spec document.
+
+### Prerequisites
+
+The following assumptions are made about the **data model** and **function registry**,
+as prerequisites for the following algorithm:
+
+1. The data model supports a message type that allows for selecting
+   one variant **case** of a message based on the value of one or more **selectors**.
+
+2. The value of a selector may be determined by a literal value,
+   a function defined in the function registry,
+   a dynamic variable provided at runtime,
+   the formatted value of another message,
+   or some subset of these.
+   The process for determining the value of any of the above is defined elsewhere.
+
+3. For each selector, a default or **fallback** value is defined.
+   This may be a fixed value such as `"other"`,
+   or a string or integer value defined separately for each selector.
+
+4. In the data model, the selectors are defined by an ordered list.
+
+5. In the data model, the variant cases are defined by an ordered list of entries.
+   In addition to the case's actual message **value**,
+   each entry contains the **key** for the case as an ordered list of string and integer values.
+   These key values are aligned with the selectors.
+   The length of any key list is not greater than the length of the selector list.
+
+6. If the function registry contains a `plural` selector function,
+   that function may be called with one number argument,
+   and be expected to return a list of string and integer values.
+   The returned values correspond to the numerical value of the argument as well as
+   the identifier of the cardinal plural category of the argument.
+
+7. The cases are sorted, such that numerical keys are before string keys and
+   default or fallback key values are after all other key values.
+   If a fallback key has a number value, it is sorted after all non-fallback keys.
+
+### Algorithm
+
+1. The value `s` of each selector is determined:
+
+   1. The initial value `v` of the selector is resolved (see prereq. 2)
+   2. If `v` is a string value or a list of string and integer values, `s = v` is set.
+   3. Else if `v` is a number and a `plural` function is defined, `s = plural(v)` is set.
```

Comment:
I'd prefer the data model to be more explicit, i.e. the selector should already encode that it will use the `plural` function for resolution.

Reply:
My intent with this is to handle cases where the data model representation does not make it clear whether the input value might or might not be a number. For instance, take the default message used by the [Fluent Playground](https://projectfluent.org/play/):

```
shared-photos =
    {$userName} {$photoCount ->
        [one] added a new photo
       *[other] added {$photoCount} new photos
    } to {$userGender ->
        [male] his stream
        [female] her stream
       *[other] their stream
    }.
```

Specifically, consider the `$photoCount` selector: at what point can it be known whether it represents a numeric value vs. e.g. the literal string `'one'`? The implicit `plural` allows us to ask that question only when we actually have the value; any earlier, we'd need additional input on its data type, or we'd need to add an explicit `plural` wrapper on every single selector.

As I mention in the discussion, the algorithm does of course allow for explicit `plural` wrappers, and that is in fact required when doing anything other than simple cardinal plural selection.